working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,510 @@
+
+<classBody> ::= // New rule.
+     '{' (<metadata> <classMemberDeclaration>)* '}'
+   | ';';
+
+<inlineClassDeclaration> ::=
+     'final'? 'inline' 'class' <classNamePart> <interfaces>? <inlineClassBody>;
+
+<inlineClassBody> ::=
+     '{' (<metadata> <inlineMemberDeclaration>)* '}'
+   | ';';
+
+<inlineMemberDeclaration> ::= <classMemberDeclaration>;
+
+<enumType> ::= // Modified rule.
+     'enum' <enumNamePart> <mixins>? <interfaces>? '{'
+        <enumEntry> (',' <enumEntry>)* (',')?
+        (';' (<metadata> <classMemberDeclaration>)*)?
+     '}';
+
+<enumNamePart> ::= // New rule.
+     <constructorName> <typeParameters>? <formalParameterList>
+   | <typeWithParameters>;
+```
+
+The word `inline` is now used in the grammar, but it is not a reserved word
+or a built-in identifier.
+
+### Static processing
+
+A class declaration with a primary constructor *(including `inline`, but
+not `<mixinApplicationClass>`, because the grammar rules do not include
+those)* is desugared to a class declaration without a primary
+constructor. An enum declaration with a primary constructor is desugared
+using the same steps. This determines the static analysis and dynamic
+semantics of a primary constructor.
+
+*In other words, there is no other semantics than the desugaring. When
+desugared, the resulting class declaration is treated exactly the same as
+it would have been if the developer had written the result of the
+desugaring step in the first place.*
+
+A compile-time error occurs if a formal parameter that contains a
+term of the form `this.v`, or `super.v` where `v` is an identifier has the
+modifier `covariant`. *For example, `required covariant int this.v` is an
+error.*
+
+Conversely, it is not an error for the modifier `covariant` to occur on
+other formal parameters of a primary constructor. *This extends the
+existing whitelist of places where `covariant` can occur.*
+
+The desugaring consists of the following steps, where _D_ is the class or
+enum declaration in the program that includes a primary constructor, and
+_D2_ is the result of desugaring. The desugaring step will delete elements
+that amount to the primary constructor, and it will add a new constructor
+_k_.
+
+Where no processing is mentioned below, _D2_ is identical to _D_. Changes
+occur as follows:
+
+The current scope of the formal parameter list of the primary constructor
+in _D_ is the current scope of the class/enum declaration *(in other words,
+the default values cannot see declarations in the class body)*.  Every
+default value in the primary constructor of _D_ is replaced by a fresh
+private name `_n`, and a constant variable named `_n` is added to the
+top-level of the current library. *(This means that we can move the
+parameter declarations including the default value without changing its
+meaning.)*
+
+Next, _k_ is a constant constructor iff the keyword `const` occurs just
+before the class name in the header of _D_, or _D_ is an `enum`
+declaration.
```

Comment:
Which means that _k_ is prefixed with the `const` modifier. Stay syntactic, we're creating new syntax here.
Saying that a constructor "exists" and "is const" at a point where the syntax of the program hasn't stabilized yet, is too soon.

(Unless you really are trying to introduce a *semantic* constructor, like the ones we infer from parsing constructor syntax, which I'd be quite happy about. But that's not what it says. I think. The distinction isn't definitely not made clear.)

Reply:
Improved now!